kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java

```diff
@@ -2130,92 +2165,106 @@ protected void handleEndTxn(KafkaHeaderAndRequest kafkaHeaderAndRequest,
+              response.complete(new WriteTxnMarkersResponse(errors));
+          }
+        };
 
-                    handleGroupFuture.whenComplete((ignored, handleGroupThrowable) -> {
-                        if (handleGroupThrowable != null) {
-                            log.error("Failed to handle group end txn for partition {}",
-                                    topicPartition, handleGroupThrowable);
-                            partitionErrorsMap.put(topicPartition, Errors.forException(handleGroupThrowable));
-                            resultFuture.complete(null);
-                            return;
-                        }
-                        String fullPartitionName = KopTopic.toString(topicPartition, namespacePrefix);
-                        TopicName topicName = TopicName.get(fullPartitionName);
-
-                        long firstOffset = transactionCoordinator.removeActivePidOffset(
-                                topicName, txnMarkerEntry.producerId());
-                        long lastStableOffset = transactionCoordinator.getLastStableOffset(topicName, offset);
-
-                        if (txnMarkerEntry.transactionResult().equals(TransactionResult.ABORT)) {
-                            transactionCoordinator.addAbortedIndex(AbortedIndexEntry.builder()
-                                    .version(request.version())
-                                    .pid(txnMarkerEntry.producerId())
-                                    .firstOffset(firstOffset)
-                                    .lastOffset(offset)
-                                    .lastStableOffset(lastStableOffset)
-                                    .build());
-                        }
-                        partitionErrorsMap.put(topicPartition, Errors.NONE);
-                        resultFuture.complete(null);
+        for (WriteTxnMarkersRequest.TxnMarkerEntry marker : markers) {
+            long producerId = marker.producerId();
+            TransactionResult transactionResult = marker.transactionResult();
+            Map<TopicPartition, MemoryRecords> controlRecords = generateTxnMarkerRecords(marker);
+            AppendRecordsContext appendRecordsContext = AppendRecordsContext.get(
+                    topicManager,
+                    requestStats,
+                    this::startSendOperationForThrottling,
+                    this::completeSendOperationForThrottling,
+                    this.pendingTopicFuturesMap);
+            getReplicaManager().appendRecords(
+                    kafkaConfig.getRequestTimeoutMs(),
+                    true,
+                    currentNamespacePrefix(),
+                    controlRecords,
+                    PartitionLog.AppendOrigin.Coordinator,
+                    appendRecordsContext
+            ).whenComplete((result, ex) -> {
+                appendRecordsContext.recycle();
+                if (ex != null) {
+                    log.error("Append txn marker failed.", ex);
```

Comment:
```suggestion
                    log.error("[{}] Append txn marker ({}) failed.", ctx.channel(), marker, ex);
```

Maybe we can also add a debug log when `ex == null`.

Reply:
Sure, Addressed.